contracts/Multisig.cairo

```diff
@@ -237,6 +237,14 @@ func require_valid_threshold{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, r
 ):
     const lower_bound = 1
 
+    # we allow threshold to be 0
+    # if and only if signers_len is 0
+    let (x) = is_not_zero(threshold)
```

Comment:
Wouldn't it be easier to just "if threshold + signers_len == 0 return ()".

Reply:
That's done because of my concern that malicious signer could execute `set_threshold` with value `-signers_len`. Which would pass all the checks and set threshold to `(Prime - 1) - signers_len`

Reply:
Hmm, interesting point. But I don't see how that applies here.
Let's say the prime is 9. If I input -1 it becomes 8 (or 7? unsure, but doesn't make a difference). The end result is the same as inputting just 8. The 'is_not_zero' only checks whether the input is zero or not (https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/cairo/common/math_cmp.cairo#L7).

Reply:
But you have a valid point about summarizing two integers. I don't think we can just `if x + y == 0:` because the other integer can be "prime - 1" and the other one can be 1.

Reply:
> If I input -1 it becomes 8 (or 7? unsure, but doesn't make a difference)
Yeah, my math was off, it is `Prime - signers_len`.

> because the other integer can be "prime - 1" and the other one can be 1.
That basically what I meant. One could choose such value of `threshold` so values would cancel out and statement `if x + y == 0` would be true.

Reply:
But I don't see how checking for zero helps. In the case of `(prime - 1) + 1` there are no zeros.

Reply:
>  In the case of (prime - 1) + 1 there are no zeros

If felt gets value  `(prime - 1) + 1` it will be 0 since we take `mod` of `prime`
Or did I misunderstand the question?

Reply:
I think I'm misunderstanding something :)
You are checking that the inputs are not zero. In our example the inputs would be `prime - 1` and `1`, and neither is zero. You are not checking the result. Therefore the checks would pass.

Reply:
Maybe there's confusion about `is_not_zero`. From cairo
`is_not_zero` -> // Returns 1 if value != 0. Returns 0 otherwise.

so on line 244 we basically check `if threshold == 0 && signers_len == 0`

Reply:
> neither is zero

Yeap so `x` and `y` would contain `1`. So on line 244 we get statement `if 2 == 0`

Reply:
Right. I took pen and paper and simulated threshold (T) and signer (S) combinations for -1, 0 and 1. From the 9 different permutations, the only ones which pass the whole function are (T0,S0), (T1,S0), (T1,S1) and (T1,S-1). The (T0,S0) is the only one which returns because of your addition, which is right how it should work. And the other cases make sense as well.

So I think it works just fine! Sorry for all the hassle. It's just very difficult for me to interpret negations such as `is_not_zero`. It's been too long since I did Cairo last time, but I guess stuff like `let (x) = threshold == 0` is not possible? Would it be clearer if the check was like `if (threshold  * threshold  + signers_len * signers_len  == 0)`?

Reply:
Yeah, that's because cairo doesn't have `is_zero`. So we basically have double negation(`!is_not_zero = !!is_zero`))

> `let (x) = threshold == 0` 


those aren't possible

> Would it be clearer if the check was like `if (threshold * threshold + signers_len * signers_len == 0)`

Well, I'm not sure which math trickery is easier to understand)

Reply:
Well you can decide. For me my version is more understandable since it doesn't include (double) negations, just pure math. Of course it'd require some commenting as well.